README.md
Plack::App::DAIA implements a DAIA webservice as PSGI application.

[![Build Status](https://travis-ci.org/nichtich/Plack-App-DAIA.png)](https://travis-ci.org/nichtich/Plack-App-DAIA)
[![Coverage Status](https://coveralls.io/repos/nichtich/Plack-App-DAIA/badge.png?branch=master)](https://coveralls.io/r/nichtich/Plack-App-DAIA?branch=master)


Changes
# Change file for Plack::App::DAIA

0.52 2013-07-31
    Use format=json in tests (fixed 0.51)
    Improved test coverage

0.51 2013-07-30
    Use format=json in tests

0.50 2013-07-11
    Added option 'safe' and removed method IDFORMAT

0.48 2013-01-10
    Added option 'errors' and method IDFORMAT
    Improved tests and removed warning

0.47 2012-06-07
    Added XML Schema

0.46 2012-05-04
    Fixed JSONP bug
    Added content negotiation
    Include XSLT client

0.45 2012-02-22
    Added initalization hook and fixed a test suite bug
    Added named capturing groups in idformat (requires Perl >= 5.10)

0.44 2012-02-21
    Moved daia_test_suite to provedaia in Plack::App::DAIA::Test::Suite

0.43 2012-01-26
    Fixed a bug and introduced another example
    Introduced idformat check and provedaia script

0.42 2012-01-18
    Introduced DAIA test suites

0.41 2012-01-17
    First draft of Plack::App::DAIA::Test
    Added turtle and HTML output to validator

0.4 2011-12-16
    Fixed DAIA/XML header bug
    Added code parameter
    Use DAIA 0.4

0.2 2011-12-14
    Added support of DAIA/RDF

0.1  2011-12-08
    First public version at CPAN


t/SampleDAIAApp.pm
package SampleDAIAApp;
use parent 'Plack::App::DAIA';

sub IDFORMAT { qr{^foo:.+} };

sub retrieve {
    my ($self, $id, %idparts) = @_;

    my $daia = DAIA::Response->new;

    $daia->addDocument( id => ($id || "foo:default") );

    # construct full response ...

    return $daia;
}

1;


lib/Plack/App/DAIA/Test.pm
use strict;
use warnings;
use v5.10.1;
package Plack::App::DAIA::Test;
#ABSTRACT: Test DAIA Servers

use base 'Test::Builder::Module';
our @EXPORT = qw(test_daia_psgi test_daia daia_app);

use URI::Escape;
use Test::More;
use Plack::Test;
use Plack::App::DAIA;
use Scalar::Util qw(reftype blessed);
use HTTP::Request::Common;
use Test::JSON::Entails;
use JSON;

sub test_daia {
    my $app = daia_app(shift) || do {
        __PACKAGE__->builder->ok(0,"Could not construct DAIA application");
        return;
    };
    my $test_name = "test_daia";
    $test_name = pop(@_) if @_ % 2;
    while (@_) {
        my $id = shift;
        my $expected = shift;
        my $res = $app->retrieve($id);
        if (!_if_daia_check( $res, $expected, $test_name )) {
            $@ = "retrieve method returned a DAIA::Response" unless $@;
            __PACKAGE__->builder->ok(0, $@);
        }
    }
}

sub test_daia_psgi {
    my $app = shift;

    # TODO: load psgi file if string given and allow for URL
    my $test_name = "test_daia";
    $test_name = pop(@_) if @_ % 2;
    while (@_) {
        my $id = shift;
        my $expected = shift;
        test_psgi $app, sub {
            my $req = shift->(GET "/?id=".uri_escape($id).'&format=xml');
            my $res = eval { DAIA::parse( $req->content ); };
            if ($@) {
                $@ =~ s/DAIA::([A-Z]+::)?[a-z_]+\(\)://ig;
                $@ =~ s/ at .* line.*//g;
                $@ =~ s/\s*$//sg;
            }
            if (!_if_daia_check( $res, $expected, $test_name )) {
                $@ = "application returned a DAIA::Response" unless $@;
                __PACKAGE__->builder->ok(0, $@);
            }
        };
    }
}

sub daia_app {
    my $app = shift;
    if ( blessed($app) and $app->isa('Plack::App::DAIA') ) {
        return $app;
    } elsif ( $app =~ qr{^https?://} ) {
        my $baseurl = $app . ($app =~ /\?/ ? '&id=' : '?id=');
        $app = sub {
            my $id = shift;
            my $url = $baseurl.$id."&format=json";
            my @daia = eval { DAIA->parse($url) };
            if (!@daia) {
                $@ ||= '';
                if ($@) {
                    $@ =~ s/DAIA::([A-Z]+::)?[a-z_]+\(\)://ig;
                    $@ =~ s/ at .* line.*//g;
                    $@ =~ s/\s*$//sg;
                }
                $@ = "invalid DAIA from $url: $@";
            }
            return $daia[0];
        };
    }
    if ( ref($app) and reftype($app) eq 'CODE' ) {
        return Plack::App::DAIA->new( code => $app );
    }
    return;
}

# Call C<$code> with C<$daia> and set as C<$_>, if C<$daia> is a L<DAIA::Response>
# and return C<$daia> on success. Return C<undef> otherwise.
sub _if_daia_check {
    my ($daia, $expected, $test_name) = @_;
    if ( blessed($daia) and $daia->isa('DAIA::Response') ) {
        if ( (reftype($expected)||'') eq 'CODE') {
            local $_ = $daia;
            $expected->($daia);
        } else {
            local $Test::Builder::Level = $Test::Builder::Level + 2;
            my $json = decode_json( $daia->json );
            entails $json, $expected, $test_name;
        }
        return $daia;
    }
}

1;

=head1 SYNOPSIS

    use Test::More;
    use Plack::App::DAIA::Test;

    use Your::App; # your subclass of Plack::App::DAIA
    my $app = Your::App->new;

    # or wrap a DAIA server
    my $app = daia_app( 'http://your.host/pathtodaia' );

    test_daia $app,
        'some:id' => sub {
            my $daia = shift; # or = $_
            my @docs = $daia->document;
            is (scalar @docs, 1, 'returned one document');
            ...
        },
        'another:id' => sub {
            my $daia = shift;
            ...
        };

    # same usage, shown here with an inline server

    test_daia_psgi
        sub {
            my $id = shift;
            my $daia = DAIA::Response->new;
            ...
            return $daia;
        },
        'some:id' => sub {
            my $daia = $_; # or shift
            ...
        };

    done_testing;

=head1 DESCRIPTION

I<This model is experimental, so take care!> The current version has different
behaviour for C<test_daia> and C<test_daia_psgi>, that might get fixed.

This module exports two methods for testing L<DAIA> servers. You must provide a
DAIA server as code reference or as instance of L<Plack::App::DAIA> and a list
of request identifiers and testing code. The testing code is passed a valid
L<DAIA::Response> object on success (C<$_> is also set to this response).

=method test_daia ( $app, $id1 => $expected, $id2 => ...  )

Calls a DAIA server C<$app>'s retrieve method with one or more identifiers,
each given a test function or an expected JSON structure to be tested with
L<Test::JSON::Entails>. This does not add warnings and the error option is
ignored (use test_daia_psgi instead if needed).

=method test_daia_psgi ( $app, $id => $expected, $id => ...  )

Calls a DAIA server C<$app> as L<PSGI> application with one or more
identifiers, each given a test function or an expected JSON structure.

=method daia_app ( $plack_app_daia | $url | $code )

Returns an instance of L<Plack::App::DAIA> or undef. Code references or URLs
are wrapped. For wrapped URLs C<$@> is set on failure. This method may be removed
to be used internally only!

=head1 SEE ALSO

L<Plack::App::DAIA::Test::Suite> and L<provedaia>.

=encoding utf8

=cut


lib/Plack/App/DAIA/Test/Suite.pm
use strict;
use warnings;
package Plack::App::DAIA::Test::Suite;
#ABSTRACT: Test DAIA Servers via a test scripting language

use base 'Test::Builder::Module';
our @EXPORT = qw(provedaia);

use Test::More;
use Plack::App::DAIA::Test;
use Scalar::Util qw(reftype blessed);
use Test::JSON::Entails;
use JSON;
use Carp;

sub provedaia {
    my ($suite, %args) = @_;

    my $test  = __PACKAGE__->builder;
    my @lines;

    if ( ref($suite) ) {
        croak 'usage: provedaia( $file | $glob | $string )'
            unless reftype($suite) eq 'GLOB' or blessed($suite) and $suite->isa('IO::File');
        @lines = <$suite>;
    } elsif ( $suite !~ qr{^https?://} and $suite !~ /[\r\n]/ ) {
        open (my $fh, '<', $suite) or croak "failed to open daia test suite $suite";
        @lines = <$fh>;
        close $fh;
    } else {
        @lines = split /\n/, $suite;
    }

    my $line = 0;
    my $comment = '';
    my $json = undef;
    my %vars = ( server => $args{server} );
    my @ids;
    @ids = @{$args{ids}} if $args{ids};

    my $run = sub {
        my $server = $vars{server} or return;
        $json ||= '{ }';
        my $server_name = $server;
        if ( $server !~ qr{^https?://}) {
            no warnings 'redefine'; # we may load the same twice
            $_ = Plack::Util::load_psgi($server);
            if ( ref($_) ) {
                diag("loaded PSGI from $server");
                $server = $_;
            } else {
                fail("failed to load PSGI from $server");
                return;
            }
        }
        foreach my $id (@ids) {
            my $test_name = "$server_name?id=$id";
            $comment =~ s/^\s+|\s+$//g;
            $test_name .= " ($comment)" if $comment ne '';
            local $Test::Builder::Level = $Test::Builder::Level + 2; # called 2 levels above
            my $test_json = $json;
            $vars{id} = $id;
            $test_json =~ s/\$([a-z]+)/defined $vars{$1} ? $vars{$1} : "\$$1"/emg;
            $test_json = decode_json($test_json);
            if (ref($server)) {
                test_daia_psgi $server, $id => $test_json, $test_name;
            } else {
                test_daia $server, $id => $test_json, $test_name;
            }
        }
    };

    foreach (@lines) {
        if ($args{end}) {
            $args{end} = 0 if /__END__/;
            next;
        }
        chomp;
        $comment = $1 if /^#(.*)/;
        s/^(#.*|\s+)$//; # empty line or comment
        $line++;

        if (defined $json) {
            $json .= $_;
            if ($_ eq '') {
                $run->();
                $json = undef;
                $comment = '';
            }
        } elsif ( $_ eq '' ) {
            next;
        } elsif( $_ =~ qr{^([a-z]+)\s*=\s*(.*)}i ) {
            $comment = '';
            my ($key, $value) = ($1,$2);
            if ($1 =~ /^id[s]?/) {
                @ids = $value eq '' ? () : ($value);
            } else {
                $vars{$key} = $value;
            }
            diag( "$key = $value" ) if $args{verbose};
        } elsif( $_ =~ qr/^\s*{/ ) {
            $json = $_;
        } else { # identifier
            $comment = '';
            push @ids, $_;
        }
    }
    $run->();
}

1;

=head1 SYNOPSIS

    use Test::More;
    use Plack::App::DAIA::Test::Suite;

    provedaia <<SUITE, server => "http://example.com/your-daia-server";
    foo:bar

    # document expected
    { "document" : [ { } ] }
    SUITE

    done_testing;

=method provedaia ( $suite [, %options ] )

Run a DAIA test suite from a string or stream (GLOB or L<IO::File>). A DAIA
test suite lists servers, identifiers, and DAIA/JSON response fragments to test
DAIA servers.  The command line client L<provedaia> is included in this
distribution for convenience.

Additional option supported so far are C<server> and C<ids>. The former is
equivalent to an inital C<server=...> statement in you test suite and the
latter is equivalent to an initial list of identifiers in you test suite.

If the option C<end> is set, all lines before C<__END__> are ignored in the
test suite script. The option C<verbose> adds more diagnostic messages.

=head1 TEST SUITE FORMAT

A test suite is defined in a text-based format that is parsed line by line.
Empty lines are ignored. There are four kinds of statements:

=over 4

=item comments

All lines starting with C<#> are treated as comments.

=item responses

All lines starting with C<{>} begin a response (fragment) in JSON format.
Following lines are treated as part of the JSON structure until an empty line
or the end of the file. References to assigned variables, such as C<$server>,
are replaced, including the special variable C<$id> for the current identifier.

=item assignements

All lines of the form C<key=value>, where C<key> contains of lowercase letters
a-z only, are treated as variable assignements. In particular, the variable
C<server> is used to set a server (an URL or a PSGI script) and the variable
C<id> can be used to reset the list of identifiers.

=item identifiers

All other non-empty lines are treated as identifiers. Identifiers are not
expected to be URI-encoded.

=back

Every time a response has been read, all preceding identifiers are used to
query the current server and the response is compared with
L<Test::JSON::Entails>. Here is an example of a test suite:

  server=http://example.com/your-daia-server

  # some document ids
  isbn:0486225437
  urn:isbn:0486225437
  http://example.org/this-is-also-an-id

  # the response must contain at least one document with the query id
  { "document" : [
    { "id" : "$id" }
  ] }

See the file C<app.psgi> and C<examples/daia-ubbielefeld.pl> for further
examples of test suites included in server implementations.

=encoding utf8

=cut


lib/Plack/App/DAIA.pm
use strict;
use warnings;
package Plack::App::DAIA;
#ABSTRACT: DAIA Server as Plack application

use v5.10.1;

use parent 'Plack::Component';
use LWP::Simple qw(get);
use Encode;
use JSON;
use DAIA;
use Scalar::Util qw(blessed);
use Try::Tiny;
use Plack::Util::Accessor qw(xslt warnings errors code idformat initialized html safe);
use Plack::Middleware::Static;
use File::ShareDir qw(dist_dir);

use Carp;
use Plack::Request;

our %FORMATS  = DAIA->formats;

sub prepare_app {
    my $self = shift;
    return if $self->initialized;

    $self->init;
    $self->errors(0) unless defined $self->errors;
    $self->warnings(1) if $self->errors or not defined $self->warnings;
    $self->idformat( qr{^.*$} ) unless defined $self->idformat;
    $self->safe(1) unless defined $self->safe;
    $self->xslt('daia.xsl') if ($self->xslt // '') =~ /^\d+$/ or (!$self->xslt && $self->html);

    if ($self->xslt) {
        if ($self->html and !ref($self->html)) {
            carp "html => 1 is deprecated, use xslt => 1 instead";
            $self->html(0);
        }
        $self->html( Plack::Middleware::Static->new(
            path => qr{daia\.(xsl|css|xsd)$|xmlverbatim\.xsl$|icons/[a-z0-9_-]+\.png$},
            root => dist_dir('Plack-App-DAIA')
        )) unless $self->html;
    }

    $self->initialized(1);
}

sub init {
    # initialization hook
}

sub call {
    my ($self, $env) = @_;
    my $req = Plack::Request->new($env);

    my $id     = $req->param('id') // '';
    my $format = lc($req->param('format') // '');

    # serve HTML client
    if ( $self->html and $id eq '' ) {
        my $resp = $self->html->_handle_static( $env );
        if ($resp and $resp->[0] eq 200) {
            return $resp;
        }
    }

    # validate identifier
    my ($invalid_id, $error, %parts) = ('',undef);
    if ( $id ne '' and ref $self->idformat ) {
        if ( ref $self->idformat eq 'Regexp' ) {
            if ( $id =~ $self->idformat ) {
                %parts = %+; # named capturing groups
            } else {
                $invalid_id = $id;
                $id = "";
            }
        }
    }

    if ( $self->warnings ) {
        if ( $invalid_id ne '' ) {
            $error = 'unknown identifier format';
        } elsif ( $id eq ''  ) {
            $error = 'please provide a document identifier';
        }
    }

    # retrieve and construct response
    my ($status, $daia) = (200, undef);
    if ( $error and $self->errors ) {
        $daia = DAIA::Response->new;
    } else {
        if ($self->safe) {
            try {
                $daia = $self->retrieve( $id, %parts );
            } catch {
                chomp($error = "request method died: $_");
                $status = 500;
            }
        } else {
            $daia = $self->retrieve( $id, %parts );
        }
        if (!$daia or !blessed $daia or !$daia->isa('DAIA::Response')) {
            $daia = DAIA::Response->new;
            $error = 'request method did not return a DAIA response'
                unless $error;
            $status = 500;
        }
    }

    if ( $error and $self->warnings ) {
        $daia->addMessage( 'en' => $error, errno => 400 );
    }

    $self->as_psgi( $status, $daia, $format, $req->param('callback') );
}

sub retrieve {
    my $self = shift;
    return $self->code ? $self->code->(@_) : undef;
}

sub as_psgi {
    my ($self, $status, $daia, $format, $callback) = @_;
    my ($content, $type);

    $type = $FORMATS{$format} unless $format eq 'xml';
    $content = $daia->serialize($format) if $type;

    if (!$content) {
        $type = "application/xml; charset=utf-8";
        if ( $self->warnings ) {
            if ( not $format ) {
                $daia->addMessage( 'en' => 'please provide an explicit parameter format=xml', 300 );
            } elsif ( $format ne 'xml' ) {
                $daia->addMessage( 'en' => 'unknown or unsupported format', 300 );
            }
        }
        $content = $daia->xml( header => 1, xmlns => 1, ( $self->xslt ? (xslt => $self->xslt) : () )  );
    } elsif ( $type =~ qr{^application/javascript} and ($callback || '') =~ /^[\w\.\[\]]+$/ ) {
        $content = "$callback($content)";
    }

    return [ $status, [ "Content-Type" => $type ], [ encode('utf8',$content) ] ];
}

1;

=head1 SYNOPSIS

Either derive from Plack::App::DAIA

    package Your::App;
    use parent 'Plack::App::DAIA';

    sub init {
        my $self = shift;
        $self->idformat( qr{^[a-z]+:.*$} ) unless $self->idformat;
    }

    sub retrieve {
        my ($self, $id, %idparts) = @_;

        my $daia = DAIA::Response->new;

        # construct full response ...

        return $daia;
    };

    1;

or pass a code reference as option C<code>:

    use Plack::App::DAIA;

    Plack::App::DAIA->new(
        code => sub {
            my ($id, %idparts) = @_;

            my $daia = DAIA::Response->new;

            # construct full response ...

            return $daia;
        },
        idformat => qr{^[a-z]+:.*$}
    );

=head1 DESCRIPTION

This module implements a B<Document Availability Information API> (L<DAIA>)
server as PSGI application. A DAIA server receives two URL parameters:

=over 4

=item B<id>

refers to the document to retrieve availability information. The id is parsed
based on the L</idformat> option and passed to an internal L</retrieve> method,
which must return a L<DAIA::Response> object.

=item B<format>

specifies a DAIA serialization format, that the resulting L<DAIA::Response> is
returned in. By default the formats C<xml> (DAIA/XML, the default), C<json>
(DAIA/JSON), and C<rdfjson> (DAIA/RDF in RDF/JSON) are supported. Additional
RDF serializations (C<rdfxml>, C<turtle>, and C<ntriples>) are supported if
L<RDF::Trine> is installed. If L<RDF::NS> is installed, the RDF/Turtle output
uses well-known namespace prefixes. Visual RDF graphs are supported with format
C<svg> and C<dot> if L<RDF::Trine::Exporter::GraphViz> is installed and C<dot>
is in C<$ENV{PATH}>.

=back

This module automatically adds appropriate warnings and error messages. A
simple HTML interface based on client side XSLT is added with option C<html>.

=method new ( [%options] )

Creates a new DAIA server. Supported options are

=over 4

=item code

Code reference to the C<retrieve> method if you prefer not to create a
module derived from this module.

=item xslt

Path of a DAIA XSLT client to attach to DAIA/XML responses (disabled by
default).  Any numeric value will be set to C<daia.xsl>, so use C<< xslt => 1
>> for enabling the default XSLT client. The default client is provided in form
of three files (C<daia.xsl>, C<daia.css>, C<xmlverbatim.xsl>) and DAIA icons,
all shipped together with this module. Enabling HTML client also enables
serving the DAIA XML Schema as C<daia.xsd>.

You may need to adjust the path if your server rewrites the request path.

=item warnings

Enable warnings in the DAIA response (enabled by default).

=item errors

Enable warnings and directly return a response without calling the retrieve
method on error.

=item idformat

Optional regular expression to validate identifiers. Invalid identifiers are
set to the empty string before they are passed to the C<retrieve> method. In
addition an error message "unknown identifier format" is added to the response,
if the option C<warnings> are enabled. If the option C<errors> is enabled,
the C<retrieve> method is not called on error.

It is recommended to use regular expressions with named capturing groups
as introduced in Perl 5.10. The named parts are also passed to the
C<retrieve method>. For instance:

  idformat => qr{^ (?<prefix>[a-z]+) : (?<local>.+) $}x

will give you C<$parts{prefix}> and C<$parts{local}> in the retrieve method.

=item safe

Catch errors on the request format if enabled (by default). You may want to
disable this to get a stack trace if the request method throws an error.

=item initialized

Stores whether the application had been initialized.

=back

=method retrieve ( $id [, %parts ] )

Must return a status and a L<DAIA::Response> object. Override this method
if you derive an application from Plack::App::DAIA. By default it either
calls the retrieve code, as passed to the constructor, or returns undef,
so a HTTP 500 error is returned.

This method is passed the original query identifier and a hash of named
capturing groups from your identifier format.

=method init

This method is called by Plack::Component::prepare_app, once before the first
request and before undefined options are set to their default values. You can
define this method in you subclass as initialization hook, for instance to set
default option values. Initialization during runtime can be triggered by
setting C<initialized> to false.

=method as_psgi ( $status, $daia [, $format [, $callback ] ] )

Serializes a L<DAIA::Response> in some DAIA serialization format (C<xml> by
default) and returns a a PSGI response with given HTTP status code.

=head1 EXAMPLES

You can also mix this application with L<Plack> middleware.

It is highly recommended to test your services! Testing is made as easy as
possible with the L<provedaia> command line script.

This module contains a dummy application C<app.psgi> and a more detailed
example C<examples/daia-ubbielefeld.pl>.

=head1 SEE ALSO

Plack::App::DAIA is derived from L<Plack::Component>. Use L<Plack::DAIA::Test>
and L<provedaia> (using L<Plack::App::DAIA::Test::Suite>) for writing tests.
See L<Plack::App::DAIA::Validator> for a DAIA validator and converter.

=encoding utf8

=cut


examples/ubbielefeld.pl
use strict;
use Plack::App::DAIA '0.42';
use DAIA;
use LWP::Simple qw(get);

#
# This script shows how to create a DAIA server with 
# Plack::App::DAIA. You can  run this server with:
#
# $ plackup ubbielefeld.pl
#

# Only use pQuery if this CPAN module is installed
our $pQuery;
BEGIN { eval { use pQuery; }; $pQuery = !$@; };


my $app = Plack::App::DAIA->new( code => \&retrieve );

sub retrieve {
    my $id = shift;
    my $daia = DAIA::Response->new;

    $daia->institution(
        content => "Universit\xE4tsbibliothek Bielefeld", # Name
        href    => "http://www.ub.uni-bielefeld.de/",     # Homepage
        id      => "http://lobid.org/organisation/DE-361" # URI (Linked Data)
    );
      
    # sort out invalid identifiers. In a future version of
    # Plack::App::DAIA this should be supported easier. 
    return if $id eq '';
    $id =~ s{^http://katalog.ub.uni-bielefeld.de/title/}{};
    if ( $id !~ /^[0-9]+$/ ) {
        $daia->addMessage( "invalid identifier format", errno => 1 );
        return $daia;
    }

    $id = 'http://katalog.ub.uni-bielefeld.de/title/' . $id;

    my $html = get($id);
    if (!$html or $@) {
        $daia->addMessage( "Failed to retrieve title data", errno => 503 );
        return $daia;
    }

    if ($html !~ /<div\s+class\s*=\s*["']\s*ex["']/ms) {
        $daia->addMessage( "Document not found", errno => 404 );
        return $daia;
    }

    my $doc = DAIA::Document->new( id => $id, href => $id );
    if ( $pQuery ) {

        # dirty old screen scraping
        pQuery( $html )->find(".ex")->find("table")->find('tr')->each(sub {
            return unless shift; # skip first row

            my $item = DAIA::Item->new;
            
            my $td = pQuery($_)->find('td');

            if ( my $standort = $td->get(0) ) {
                if ( $standort->find('a') ) {
                    $standort = $standort->firstChild;
                    my $url = $standort->getAttribute('href');
                    $item->department( "".$standort->innerHTML, href => $url );
                } else {
                    $item->department( "".$standort->innerHTML );
                }
            }
            if ( my $storage = $td->get(1) ) {
                $item->storage( "".$storage->innerHTML );
            }
            if ( my $label = $td->get(2) ) {
                $item->label( "".$label->innerHTML );
            }

            my $loan   = eval { $td->get(4)->innerHTML; } || "";
            my $status = eval { $td->get(5)->innerHTML; } || "";
            my $queue  = eval { 1 * $td->get(6)->innerHTML; } || 0;
            
            if ( $status =~ /Entliehen bis (\d\d)\.(\d\d)\.(\d\d\d\d)/ ) {
                $item->addUnavailable( service => "loan", expected => "$3-$2-$1", href => $id );
                $item->addUnavailable( service => "presentation", expected => "$3-$2-$1" );
            } elsif( $loan =~ /Nicht ausleihbar/ ) {
                $item->addUnavailable( service => "loan" );
            } else {
                # probably you can get it (fuzzy heruistic)
                $item->addAvailable( service => "loan" );
                $item->addAvailable( service => "presentation" );
            }

            $doc->addItem($item);
        });
    }

    $daia->document( $doc );

    return $daia;
}

$app;

__END__
#
# You can test the running server with Plack::App::DAIA::Test via:
#
# awk 'p {print} /__END__/ {p=1}' ubbielefeld.pl | \
#   perl -MPlack::App::DAIA::Test -MTest::More -e 'daia_test_suite \*STDIN; done_testing;'
#

server=http://localhost:5000/

# TODO: support id=http://katalog.ub.uni-bielefeld.de/title/218777
218777

# found at least one item
{ "document" : [ {
    "id" : "http://katalog.ub.uni-bielefeld.de/title/218777",
    "item": [ { } ]
} ]}


lib/Plack/App/DAIA/Validator.pm
use strict;
use warnings;
package Plack::App::DAIA::Validator;
#ABSTRACT: DAIA validator and converter

use CGI qw(:standard);
use Encode;
use File::ShareDir qw(dist_dir);
use File::Spec::Functions qw(catfile);

use parent 'Plack::App::DAIA';
use Plack::Util::Accessor qw(xsd xslt warnings html);

our $HAS_LIBXML = 0;

our ($FORMATS);
BEGIN {
    my %f = DAIA->formats;
    $FORMATS = { map { $_ => $_ } keys %f };
    $FORMATS->{html}     = 'DAIA/HTML';
    $FORMATS->{json}     = 'DAIA/JSON';
    $FORMATS->{xml}      = 'DAIA/XML';
    $FORMATS->{rdfjson}  = 'DAIA/RDF (JSON)';
    $FORMATS->{turtle}   = 'DAIA/RDF (Turtle)'   if $FORMATS->{turtle};
    $FORMATS->{ntriples} = 'DAIA/RDF (NTriples)' if $FORMATS->{ntriples};
    $FORMATS->{rdfxml}   = 'DAIA/RDF (RDF/XML)'  if $FORMATS->{rdfxml};
    foreach (qw(dot svg)) {
        $FORMATS->{$_} = "DAIA/RDF graph ($_)" if $FORMATS->{$_};
    }
}

sub init {
    my $self = shift;
    if ($self->xsd) {
        ## no critic
        eval "use XML::LibXML";
        $HAS_LIBXML = !$@;
    }
}

sub call {
    my ($self, $env) = @_;
    my $req = Plack::Request->new($env);

    my $msg = "";
    my $error = "";
    my $url  = $req->param('url') || '';
    my $data = $req->param('data') || '';
    #eval{ $data = Encode::decode_utf8( $data ); }; # icoming raw data is UTF-8

    my $eurl = $url; # url_encode
    $eurl =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;

    my $xsd = $self->xsd;

    my $informat  = lc($req->param('in'));
    my $outformat = lc($req->param('out')) || lc($req->param('format')) || 'html';

    my $callback  = $req->param('callback') || "";
    $callback = "" unless $callback =~ /^[a-z][a-z0-9._\[\]]*$/i;

    my @daiaobjs;

    # parse DAIA
    if ( $data ) {
        @daiaobjs = eval { DAIA->parse( data => $data, format => $informat ) };
    } elsif( $url ) {
        @daiaobjs = eval { DAIA->parse( file => $url, format => $informat ) };
    }
    if ($@) {
        $error = $@;
        $error =~ s/DAIA::([A-Z]+::)?[a-z_]+\(\):| at .* line.*//ig;
    }

    my $daia;
    if (@daiaobjs > 1) {
        $error = "Found multiple DAIA elements (".(scalar @daiaobjs)."), but expected one";
    } elsif (@daiaobjs) {
        $daia = shift @daiaobjs;
    }

    if ( $FORMATS->{$outformat} and $outformat ne 'html' ) {
        $daia = DAIA::Response->new() unless $daia;
        $daia->addMessage( $error, errno => 500, lang => 'en' ) if $error;
        return $self->as_psgi( 200, $daia, $outformat, $req->param('callback') );
    } elsif ( $outformat ne 'html' ) {
        $error = "Unknown output format - using HTML instead";
    }

    # HTML output
    $error = "<div class='error'>".escapeHTML($error)."!</div>" if $error;
    if ( $url and not $data ) {
        $msg = "Data was fetched from URL " . a({href=>$url},escapeHTML($url));
        $msg .= " (" . a({href=>'#result'}, "result...") . ")" if $daia;
        $msg =  div({class=>'msg'},$msg);
#        $msg .= div({class=>'msg'},"Use ".
#                    a({href=>url()."?url=$eurl"},'this URL') .
#                    " to to directly pass the URL to this script.");
    }

    my $html = <<HTML;
<html>
<head>
  <title>DAIA Validator</title>
  <meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
  <style>
    body { font-family: arial, sans-serif;}
    h1, p { margin: 0; text-align: center; }
    h2 { margin-top: 2px; border-bottom: 1px dotted #666;}
    form { margin: 1em; border: 1px solid #333; }
    fieldset { border: 1px solid #fff; }
    label, .error, .msg { font-weight: bold; }
    .submit, .error { font-size: 120%; }
    .error { color: #A00; margin: 1em; }
    .ok { color: #0A0; margin: 1em; }
    .msg { color: #0A0; margin: 1em; }
    .footer { font-size: small; margin: 1em; }
    #result { border: 1px dotted #666; margin: 1em; padding: 0.5em; }
  </style>
</head>
<body>
<h1 id='top'>DAIA Validator/Converter</h1>
<p>Convert and Validate <a href="http://purl.org/NET/DAIA">DAIA response format</a></p>
<form method="post" accept-charset="utf-8" action="">
HTML

    $html .= $msg . $error .
     fieldset(label('Input: ',
            popup_menu('in',['','json','xml'],'',
                       {''=>'Guess','json'=>'DAIA/JSON','xml'=>'DAIA/XML'})
      )).
      fieldset('either', label('URL: ', textfield(-name=>'url', -size=>70, -value => $url)),
        'or', label('Data:'),
        textarea( -name=>'data', -rows=>20, -cols=>80, -value => $data),
      ).
      fieldset(
        label('Output: ',
            popup_menu('out',
                [ sort { $FORMATS->{$a} cmp $FORMATS->{$b} } keys %$FORMATS ],
                $outformat, $FORMATS )
        ), '&#xA0;',
        label('JSONP Callback: ', textfield(-name=>'callback',-value=>$callback))
      ).
      fieldset('<input type="submit" value="Convert" class="submit" />')
    ;
    my $has_graphviz = grep /^(svg|dot)$/, keys %$FORMATS;
    if ( $has_graphviz && $url && !$data) {
       $html .= "<fieldset>See RDF graph <a href=\"?url=$eurl&format=svg\">as SVG</a></fieldset>";
    }
    $html .= '</form>';

    if ($daia) {
      if ( $informat eq 'xml' or DAIA::guess($data) eq 'xml' ) {
        # TODO: move this into module DAIA (validate option when parsing)
        my ($schema, $parser);
        if ($xsd) {
            if (!$HAS_LIBXML) {
                $error = "XML::LibXML not found - validating skipped";
            } else {
                $parser = XML::LibXML->new;
                $schema = eval {
                    XML::LibXML::Schema->new(
                        location => catfile(dist_dir('Plack-App-DAIA'),'daia.xsd')
                    );
                };
                if ($schema) {
                    my $doc = $parser->parse_string( $data );
                    eval { $schema->validate($doc) };
                    $error = "DAIA/XML not valid but parseable: " . $@ if $@;
                } else {
                    $error = "Could not load XML Schema - validating skipped";
                }
            }
        }
        if ( $error ) {
          $html .= "<p class='error'>".escapeHTML($error)."</p>";
        } else {
          $html .= "<p class='ok'>DAIA/XML valid according to ".a({href=>$xsd},"this XML Schema")."</p>";
        }
      } else {
         $html .= p("validation is rather lax so the input may be invalid - but it was parseable");
      }
      $html .= "<div id='result'>";
      my ($pjson, $pxml, $pttl) = ("","","");
      if (!$data && $url) {
        $pjson = $pxml = $pttl = "?callback=$callback&url=$eurl";
        $pjson = " (<a href='$pjson&format=json'>get via proxy</a>)";
        $pxml  = " (<a href='$pxml&format=xml'>get via proxy</a>)";
        #$pttl  = " (<a href='$pttl&format=turtle'>get via proxy</a>)";
      }
      $html .= "<h2 id='json'>Result in DAIA/JSON$pjson <a href='#top'>&#x2191;</a> <a href='#xml'>&#x2193;</a></h2>";
      $html .= pre(escapeHTML( encode('utf8',$daia->json( $callback ) )));
      $html .= "<h2 id='xml'>Result in DAIA/XML$pxml <a href='#json'>&#x2191;</a></h2>";
      $html .= pre(escapeHTML( encode('utf8',$daia->xml( xmlns => 1 ) )));
      if ($FORMATS->{turtle}) {
        $html .= "<h2 id='ttl'>Result in DAIA/RDF (Turtle) <a href='#json'>&#x2191;</a></h2>";
        my $ttl = $daia->serialize('turtle');
        $html .= pre(escapeHTML( encode('utf8', $ttl )));
      }
      $html .= "</div>";
    }

    $html .= "<div class='footer'>Based on ";
    $html .= join ' and ', map {
        "<a href='http://search.cpan.org/perldoc?$_'>$_</a> " . ($_->VERSION || '');
    } qw(Plack::App::DAIA::Validator DAIA);
    $html .= <<HTML;
. Visit the <a href="http://github.com/gbv/daia/">DAIA project at github</a> for sources and details.
</div></body>
HTML

    return [ 200, [ 'Content-Type' => 'text/html; charset=utf-8' ], [ $html ] ];
}

1;

=head1 SYNOPSIS

    use Plack::Builder;
    use Plack::App::DAIA::Validator;

    builder {
        enable 'JSONP';
        Plack::App::DAIA::Validator->new(
            xsd      => $location_of_daia_xsd,
            html     => 1,
            xslt     => "/daia.xsl",
            warnings => 1
        );
    };

=head1 DESCRIPTION

This module provides a simple L<DAIA> validator and converter as PSGI web
application.

To support fetching from DAIA Servers via HTTPS you might need to install
L<LWP::Protocol::https> version 6.02 or higher.

=head1 CONFIGURATION

See L<Plack::App::DAIA> for documentation of options C<xslt>, C<html>,
C<warnings>, and C<xsd>. L<XML::LibXML> must be installed if C<xsd> is set to
validate DAIA/XML.

=encoding utf8

=cut
